packages/mdc-button/_mixins.scss

```diff
@@ -92,48 +186,63 @@
   border-width: $outline-width;
 }
 
-@mixin mdc-button-base_() {
-  @include mdc-typography(button);
-  @include mdc-ripple-surface;
-  @include mdc-ripple-radius-bounded;
-  @include mdc-button-horizontal-padding($mdc-button-horizontal-padding);
-
-  display: inline-flex;
-  position: relative;
-  align-items: center;
-  justify-content: center;
-  box-sizing: border-box;
-  min-width: 64px;
-  height: $mdc-button-height;
-  border: none;
-  outline: none;
-  /* @alternate */
-  line-height: inherit;
-  user-select: none;
-  -webkit-appearance: none;
-  overflow: hidden;
-  vertical-align: middle;
-
-  &::-moz-focus-inner {
-    padding: 0;
-    border: 0;
-  }
-
-  // postcss-bem-linter: ignore
-  &:active {
-    outline: none;
+@mixin mdc-button-base_($query) {
+  $feat-color: mdc-feature-create-target($query, color);
+  $feat-structure: mdc-feature-create-target($query, structure);
+  $feat-typography: mdc-feature-create-target($query, typography);
+
+  @include mdc-feature-targets($feat-typography) {
+    @include mdc-typography(button);
   }
 
-  &:hover {
-    cursor: pointer;
+  @include mdc-ripple-surface($query);
+
+  @include mdc-feature-targets($feat-structure) {
+    @include mdc-ripple-radius-bounded;
+    @include mdc-button-horizontal-padding($mdc-button-horizontal-padding);
+
+    display: inline-flex;
+    position: relative;
+    align-items: center;
+    justify-content: center;
+    box-sizing: border-box;
+    min-width: 64px;
+    height: $mdc-button-height;
+    border: none;
+    outline: none;
+    /* @alternate */
+    line-height: inherit;
+    user-select: none;
+    -webkit-appearance: none;
+    overflow: hidden;
+    vertical-align: middle;
+
+    &::-moz-focus-inner {
+      padding: 0;
+      border: 0;
+    }
+
+    // postcss-bem-linter: ignore
+    &:active {
+      outline: none;
+    }
+
+    &:hover {
+      cursor: pointer;
+    }
   }
 
   &:disabled {
```

Comment:
Curious, wouldn't the below test fail when all the feature are turned off and this emits empty style rule? Unless we're looking for only the existence of CSS properties.

```
$:disabled {
}
```

Reply:
Matt also asked this question. Sass is smart enough to not emit the containing selector if there are no rules. See https://github.com/material-components/material-components-web/pull/4228#discussion_r247692856